### PR TITLE
[5.1] Introduce CheckHttpCache middleware

### PR DESCRIPTION
**Pull request by rkgrep**
Provides native symfony handling of cached responses.

I propose include this in 5.1 and enable by default in 5.2 base project.

### CODE REVIEW

**Comment by rkgrep** [comment]
P. S. This fixed intervention/imagecache behavior, which [adds proper cache info into response](https://github.com/Intervention/imagecache/blob/master/src/Intervention/Image/ImageCacheController.php#L153), but it is completely ignored by Laravel.

**Comment by GrahamCampbell** [comment]
Not sure if this really belongs in the core? // cc @taylorotwell

**Comment by rkgrep** [comment]
I'm not sure if two lines of code worth creating a package, but I have to implement this almost in every project that uses JSON output or especially imagecache package.

So I think it would be quite handy at least to have it in Foundation.

**Comment by GrahamCampbell** [comment]
This is small enough that I'd recommend literally copy/pasting it into any app you want to use.

**Comment by GrahamCampbell** [comment]
It's basically 3 lines of actual code.

**Comment by rkgrep** [comment]
#6341 is different. It included E-tag by default, which is not the same.
This one just checks - if the response already contains e-tag, last-modified etc. and compares it with request. If everything matches, it returns 304 and empty response.

**Comment by taylorotwell** [comment]
Is there any point in even having a middleware for this? Should we do this at the same time we do $response->prepare in the core?

**Comment by rkgrep** [comment]
@taylorotwell I thought about that. Middleware gives more control as it can be disabled or applied to specific routes.
But it is up to you to decide.

**Comment by lucasmichot** [comment]
@taylorotwell this is how it's done in `dingo/api` package: https://github.com/dingo/api/blob/master/src/Routing/Router.php

**Comment by rkgrep** [comment]
Dingo also automaticaly adds etag and allows to disable it in route parameters or in config. However it is made for APIs which return JSON. HTML must not be passed to `md5` each time IMHO. This will be useles if you have data that changes on each page update.

**Comment by rkgrep** [comment]
Also there are cases when you may want to return 304 even before content is rendered.

**Comment by taylorotwell** [comment]
OK so can someone explain in plain english why we would even want a middleware for this and wouldn't just bake it into the router?

**Comment by rkgrep** [comment]
1. It can be applied only to specific routes.
2. It can be disabled globally with `middleware.disable`.
3. It can be extended in any way by creating a descendant class.
4. It can be included in 5.1 without breaking changes.

**Comment by rkgrep** [comment]
I would only agree with idea to bake it into router/kernel in 5.3 - so it can properly tested on dev projects and maybe somehow improved.

**Comment by GrahamCampbell** [comment]
I don't think baking it into the router is a good idea, just like it isn't for sessions. A middleware feels like a natural way to deal with this.

**Comment by taylorotwell** [comment]
Well if anything the class names needs to change. It's not descriptive at all in terms of what is actually happening.

**Comment by taylorotwell** [comment]
Renamed to `CheckResponseForModifications` and moved to `illuminate/http` instead of tying it up in Foundation.

**Comment by rkgrep** [comment]
:+1:
